### DISCUSSION TITLE
Version 2.4 RC 1

### DISCUSSION THREAD

**Discussion by vkbo** [vkbo/novelWriter#1798] [General]
### Release Notes

This is a release candidate of the next release version, and is intended for testing purposes. Please be careful when using this version on writing projects, and make frequent backups.

### Detailed Changelog

**Bugfixes**

* A number of issues GUI icon scaling has been fixed. Icons and buttons are supposed to scale relative to the GUI font size, but several of them were not. Issue #1787. PR #1788.
* The persistence of button state for the editor search box has been improved. The present state was only saved when the search box was actively closed, and there were several scenarios where this didn't happen. The button states are now tracked in the central config, and will always be preserved. Issue #1794. PR #1795.

**Other Improvements**

* A placeholder icon has been added for cases where users have their own icon theme, and icons are missing. Issue #1780. PR #1781.
* The project search now refreshes the search results for the currently open document when it's being edited. PR #1782.
* When activating project search from the editor while text is selected, the search box is populated with this text. Issue #1789. PR #1790.

**Packaging**

* The project has moved from using a `setup.cfg` to only using the `pyproject.toml` meta data file for packaging. This is the preferred file format now. PR #1791.

**Code Improvements**

* The part of the code related to Qt widgets and flags has been updated and refactored in preparation for the eventual move to the Qt6 framework. Part of issue #1142. PR #1792.
* The Open Document writer class has been refactored and improved. PR #1796.

<hr /><em>This discussion was created from the release <a href='https://github.com/vkbo/novelWriter/releases/tag/v2.4rc1'>Version 2.4 RC 1</a>.</em>

**Comment by ahoythedinghy** [comment]
You are feature adding essentially the only things I've wished for in novelWriter. Thank you for your work!

**Comment by awqk** [comment]
Global search update still has a small problem and does not reflect all document changes.
I think it's missing changes when editing while the search update is running.

**Reply by vkbo** [reply]
The search result for the currently open document is not updated in realtime. That's too much processing. It currently runs on the same timer as the word counter, which is every 5 seconds when the document is actively edited. It is not guaranteed to be up to date at any given time.

I can add a trigger that also updates when a document is saved. But this will never really be perfect.

**Reply by awqk** [reply]
Something should be possible, like retrigger the update if anything happened during updating, or mark the search result as dirty, due for another update. If it is too complex to do, it might not be worth the effort.

Edit: Don't Auto-Update at all, add a reload button?

**Reply by vkbo** [reply]
I'm not sure the problem is that the document changes during editing. The search is not performed on the document itself, but on a snapshot of the text. What issues exactly are you seeing? If you need to use the search tool for rapid direct editing, the in-document search tool is better suited for it.

Anyway, it's not a matter of complexity. The alternative here is to trigger an update on each keystroke, or on a very short timer. That will create noticeable skips and jerks in the response time of the UI. I can move some of the processing to a different thread, but the document itself has to be in the main thread (it's a Qt restriction) so reading the text buffer will still be in the same thread as your typing actions.

The other alternative is to not do an update from text changes at all and instead rerun the search when you interact with the search result. This is trickier because the results are deleted and added back again, so when you click a result, it is difficult to re-select it after the search is updated.

**Reply by awqk** [reply]
Update on every keystroke seems excessive. You could mark dirty on every keystroke, and rerun on timer or after some idle period.

Perhaps my problem here is that the search results sometimes are accurate, sometimes they aren't.

Thinking about another editor I regularly use, the search result list never updates. I soon learned to edit bottom-up to keep the locations accurate as long as possible. When needed, I rerun the search manually.

**Reply by awqk** [reply]
> What issues exactly are you seeing?

I managed to get the panel to a state where it wouldn't update the search result for one of the files at all (not automatically, that is).

**Reply by vkbo** [reply]
> Thinking about another editor I regularly use, the search result list never updates. I soon learned to edit bottom-up to keep the locations accurate as long as possible. When needed, I rerun the search manually.

That was the original implementation, but I found that a little annoying. I pushed a [single commit](https://github.com/vkbo/novelWriter/commit/6b9c81e8c48c87af18d6381d3981c6714f571497) yesterday that will update the search automatically if you press `Ctrl+S` to save the document.

> I managed to get the panel to a state where it wouldn't update the search result for one of the files at all (not automatically, that is).

Hmm, I suspected something here, and I tested. It seems the last document in the list doesn't get updated. Can you confirm this is the issue you're seeing when not getting updates?

The way the result is updated is that the old result is deleted, and a new inserted at the same index in the search result tree. There is some boundary checks for index 0 and last index, so there may be an off by one issue here. I did fix it for index 0 I believe.

**Reply by awqk** [reply]
like this?
added "test1" at end of a document (not the last document btw)
search global "test1" --> 1 hit (this is both the first and the last hit naturally)
changed "test1" to "test2"
search still shows 1 hit, no update (click navigates to "test2")
added "test1" to document, after "test2"
search updates, 1 hit (click navigates to "test1")

If there are two hits and I change one of them, the hits update correctly.

edit:
no hits + add 3 hits to document --> no update
3 hits + remove all hits --> no update

**Reply by vkbo** [reply]
I'm wondering if your "mark contents dirty" approach is a better option. I can flag the specific document as dirty, and re-run the search on that specific document when you interact with the result.

The question is under what criteria should the search be run again? Say you click the 5th result out of 10. Should it re-run the search then? When the search has been re-run, that result may have moved or been removed. What would you expect the UI to do then? Open the result at the same location in the list (the 5th one) or just clear the selection?

**Reply by awqk** [reply]
Assuming you have the "mark dirty" part covered, your question is when to update. All results in a document are possibly invalid. Options coming to mind:

- like you said, but it is cumbersome/impossible to maintain a sensible position in the result list. Sub-options:
  - ask the user if he wants to refresh the search (I would probably say no, because from past experience I would have started editing back to front)
  - refresh without asking (I would hate this, because I need to re-click the result once the refresh is done)
- refresh after some idle time (no user interaction)
  - refresh will also happen at inopportune times
- never, do it manually (dirty status could be shown somewhere, for example activate/deactivate search button, or within each result)
  - this would be my preference

There must be scores of editor/IDE designers who already struggled with the same question. Unfortunately I can't come up with the right search terms to tap their wisdom.

**Reply by vkbo** [reply]
My go-to for ideas for these kind of features is VSCode. The search feature I implemented is based on that, although significantly simpler. In VSCode, the search result is updated in (perceptible) real time. As mentioned, this is not so trivial here since it will be noticeably laggy.

Another point is that I would prefer to not initiate a search refresh from the editor at all, like I do now. The search widget should make that determination. Emitting a dirty flag is trivial, so that doesn't count in this context.

I any case, I tend towards your third option as well. I could highlight the results that are dirty by changing background colour, and offer an option to refresh that particular document. Perhaps by revealing a button somewhere, and also offering it on right-click.

**Reply by awqk** [reply]
Yes, VSCode gets it right, I never consciously noticed because I didn't expect it to work.

**Reply by awqk** [reply]
They text of the displayed matches is sometimes misleading. Please consider adding some preceding context, like VSCode does.
search john --> 
"John are back in a sequel to My Novel!" --> "...Jane and John are back in a sequel..."

**Reply by vkbo** [reply]
The problem is that the text cannot be highlighted like VSCode does, so it gets difficult to see where the search result is. I can try to extend it to include the full word at least.

**Reply by awqk** [reply]
Ah, you mean the highlighting in the results panel. I see the problem.

- Add the preceding word as well (unless the matching word starts a sentence)? This might give some better sense of the location.
"John are back in a sequel to My Novel!" --> "...and John are back in a sequel to..."
- choose a different panel class, if there is one, supporting rich text?

**Reply by vkbo** [reply]
> choose a different panel class, if there is one, supporting rich text?

If there was one, I would have already. My only option is to implement my own paint function to draw the different colours. It isn't difficult, I have plenty of those in novelWriter already, but I think performance would be an issue here due to the sheer number of them (potentially). If I was writing C++, instead of Python wrapping C++, I wouldn't hesitate to draw my own widget from scratch.

The welcome dialog project list is custom drawn item widgets, but it's not likely to be such a long list.

In theory, you can have 1000 search results per document in your project. I can of course test it and see if it's a problem. I can also cap the max results to a smaller number, but it can still be pretty big.

**Reply by awqk** [reply]
I'm no longer sure what the status is on this. Last time I checked (2.4.3) there are no automatic updates of the search results any more. Correct?

**Reply by vkbo** [reply]
There are no changes to search in the 2.4.x patches after release. I only do fixes in patches.

**Reply by vkbo** [reply]
I added #1830 to the 2.5 release milestone.
